### Antoni Radomiński-Lasek

In [49]:
import psycopg2 as pg
import pandas as pd
from sqlalchemy import create_engine

In [50]:
# Połączenie za pomocą psycopg2 i pandas
connection = pg.connect(host='pgsql-196447.vipserv.org', port='5432', dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

In [51]:
# Łączenie się za pomocą sqlAlchemy i pandas

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"

db = create_engine(db_string)
connection_sqlalchemy = db.connect()

ZADANIE 1.  
Ile kategorii filmów mamy w wypożyczlni?

In [52]:
df = pd.read_sql('select category_id from category',con=connection_sqlalchemy)
df.size

16

Odpowiedź:  
Mamy 16 kategorii filmów.

ZADANIE 2.   
Wyświetl listę kategorii w kolejności alfabetycznej.


In [53]:
df = pd.read_sql('SELECT name FROM category ORDER BY name ASC;',con=connection_sqlalchemy)
df


name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

ZADANIE 3.  
Znajdź najstarszy i najmłodszy film do wypożyczenia.

In [54]:
df = pd.read_sql('SELECT DISTINCT release_year FROM film',con=connection_sqlalchemy)
df

release_year
0          2006

Wszystkie filmy są z 2006 roku.

ZADANIE 4.  
Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01

In [55]:
df = pd.read_sql("SELECT rental_date FROM rental WHERE rental_date BETWEEN '2005-07-01' AND '2005-08-01'",con=connection_sqlalchemy)
df.size

6709

ODP. Wypożyczono 6709 filmów.

ZADANIE 5.  
Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?

In [56]:
df = pd.read_sql("SELECT rental_date FROM rental WHERE rental_date BETWEEN '2010-01-01' AND '2011-02-01'",con=connection_sqlalchemy)
df.size

0

ODP: Nie został wypożyczony żaden film.

ZADANIE 6.  
Znajdź największą płatność wypożyczenia.

In [57]:
df = pd.read_sql("SELECT amount FROM payment ORDER BY amount DESC LIMIT 1;", con=connection_sqlalchemy)
df

amount
0   11.99

ODP. Dość na około ale pozwala to wykożystać instrukcje select z ćwiczenia. Największa płatność to 11,99.

ZADANIE 7.  
Znajdź wszystkich klientów z Polski, Nigerii i Bangladeszu.

In [58]:
df = pd.read_sql("""SELECT c.first_name, c.last_name, p.country 
                    FROM customer c
                    INNER JOIN address a ON c.address_id = a.address_id
                    INNER JOIN city m ON a. city_id = m.city_id
                    INNER JOIN country p ON m.country_id = p.country_id
                    WHERE country IN ('Poland', 'Nigeria', 'Bangladesh')
                    ORDER BY p.country""",con=connection_sqlalchemy)

df

first_name  last_name     country
0    Michelle      Clark  Bangladesh
1     Stephen     Qualls  Bangladesh
2       Frank   Waggoner  Bangladesh
3     Marilyn       Ross     Nigeria
4       Elsie     Kelley     Nigeria
5      Gladys   Hamilton     Nigeria
6       Sonia    Gregory     Nigeria
7      Rodney    Moeller     Nigeria
8       Velma      Lucas     Nigeria
9       Carol     Garcia     Nigeria
10       Olga    Jimenez     Nigeria
11     Bertha   Ferguson     Nigeria
12     Tracey    Barrett     Nigeria
13         Jo     Fowler     Nigeria
14    Wallace      Slone     Nigeria
15  Constance       Reid     Nigeria
16      Brian      Wyman      Poland
17     Sidney   Burleson      Poland
18   Marjorie     Tucker      Poland
19    Russell    Brinson      Poland
20       Leah     Curtis      Poland
21      Ruben      Geary      Poland
22    Johnnie   Chisholm      Poland
23     Jimmie  Eggleston      Poland

ODP. Widoczna tabela prezentuje połączenie dwóch innych tabel za pomocą funkcji INNER JOIN.

ZADANIE 8.  
Gdzie mieszkają członkowie personelu?

In [59]:
df = pd.read_sql("""SELECT s.first_name, s.last_name, a.address, m.city, p.country 
                    FROM staff s
                    INNER JOIN address a ON s.address_id = a.address_id
                    INNER JOIN city m ON a. city_id = m.city_id
                    INNER JOIN country p ON m.country_id = p.country_id
                    ORDER BY s.last_name""", con = connection_sqlalchemy)
df

first_name last_name               address        city    country
0       Mike   Hillyer     23 Workhaven Lane  Lethbridge     Canada
1        Jon  Stephens  1411 Lillydale Drive   Woodridge  Australia

ZADANIE 9.  
Ilu pracowników mieszka w Argentynie lub Hiszpani?

In [60]:
df = pd.read_sql("""SELECT p.country 
                    FROM staff s
                    INNER JOIN address a ON s.address_id = a.address_id
                    INNER JOIN city m ON a. city_id = m.city_id
                    INNER JOIN country p ON m.country_id = p.country_id
                    WHERE p.country IN ('Argentina','Spain')""", con = connection_sqlalchemy)
df.size

0

ODP. Żaden pracownik nie mieszka w Argentynie lub hiszpani.

ZADANIE 10.  
Jakie kategorie filmów zostały wypożyczone przez klientów?

In [61]:
df = pd.read_sql("""SELECT DISTINCT cat.name
                    FROM customer c
                    INNER JOIN rental r ON c.customer_id = r.customer_id
                    INNER JOIN inventory i ON r.inventory_id = i.inventory_id
                    INNER JOIN film_category f ON i.film_id = f.film_id
                    INNER JOIN category cat ON f.category_id = cat.category_id
                    ORDER BY cat.name""", con = connection_sqlalchemy)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

ZADANIE 11.  
Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.

In [62]:
df = pd.read_sql("""SELECT DISTINCT cat.name
                    FROM customer c
                    INNER JOIN rental r ON c.customer_id = r.customer_id
                    INNER JOIN inventory i ON r.inventory_id = i.inventory_id
                    INNER JOIN film_category f ON i.film_id = f.film_id
                    INNER JOIN category cat ON f.category_id = cat.category_id
                    
                    INNER JOIN address ON c.address_id = address.address_id
                    INNER JOIN city ON address.city_id = city.city_id
                    INNER JOIN country ON city.country_id = country.country_id
                    WHERE country.country IN ('United States')
                    ORDER BY cat.name""", con = connection_sqlalchemy)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

ZADANIE 12.  
Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley.

In [63]:
df = pd.read_sql("""SELECT a.first_name, a.last_name, f.title
                    FROM film f
                    INNER JOIN film_actor fa ON fa.film_id = f.film_id
                    INNER JOIN actor a ON fa.actor_id = a.actor_id
                    WHERE a.first_name = 'Olympia' AND a.last_name = 'Pfeiffer' OR
                    a.first_name = 'Julia' AND a.last_name = 'Zellweger' OR
                    a.first_name = 'Ellen' AND a.last_name = 'Presley'
                    ORDER BY f.title""", con = connection_sqlalchemy)
df

first_name  last_name                   title
0     Olympia   Pfeiffer             Badman Dawn
1       Ellen    Presley         Bilko Anonymous
2       Julia  Zellweger    Breakfast Goldfinger
3       Ellen    Presley       Caribbean Liberty
4       Ellen    Presley        Casper Dragonfly
5     Olympia   Pfeiffer             Chitty Lock
6     Olympia   Pfeiffer      Color Philadelphia
7     Olympia   Pfeiffer       Contact Anonymous
8       Julia  Zellweger        Cranes Reservoir
9       Julia  Zellweger             Dares Pluto
10    Olympia   Pfeiffer            Deep Crusade
11      Julia  Zellweger        Detective Vision
12      Julia  Zellweger         Divorce Shining
13    Olympia   Pfeiffer        Effect Gladiator
14      Ellen    Presley        Empire Malkovich
15    Olympia   Pfeiffer          Express Lonely
16    Olympia   Pfeiffer       Firehouse Vietnam
17      Ellen    Presley           Floats Garden
18      Ellen    Presley        Frogmen Breaking
19    Olympia   Pfeiffer        Fugitive Maguire
20    Olympia   Pfeiffer           Hanky October
21      Julia  Zellweger         Hollow Jeopardy
22      Ellen    Presley          Homeward Cider
23      Ellen    Presley             Hyde Doctor
24    Olympia   Pfeiffer            Ice Crossing
25    Olympia   Pfeiffer         Idols Snatchers
26      Ellen    Presley          Image Princess
27    Olympia   Pfeiffer  Intolerable Intentions
28      Ellen    Presley           Jacket Frisco
29      Julia  Zellweger         Jeopardy Encino
30      Julia  Zellweger        Lambs Cincinatti
31    Olympia   Pfeiffer      Magnolia Forrester
32      Julia  Zellweger         Majestic Floats
33    Olympia   Pfeiffer              Mars Roman
34    Olympia   Pfeiffer               Maude Mod
35      Ellen    Presley    Microcosmos Paradise
36      Julia  Zellweger            Minds Truman
37    Olympia   Pfeiffer        Murder Antitrust
38      Ellen    Presley            Network Peak
39    Olympia   Pfeiffer            None Spiking
40      Julia  Zellweger            Open African
41      Ellen    Presley              Oscar Gold
42    Olympia   Pfeiffer             Others Soup
43      Julia  Zellweger            Outlaw Hanky
44      Julia  Zellweger         Panky Submarine
45      Ellen    Presley          Pickup Driving
46      Ellen    Presley         Pinocchio Simon
47      Ellen    Presley            Private Drop
48    Olympia   Pfeiffer           Psycho Shrunk
49      Julia  Zellweger        Rider Caddyshack
50      Ellen    Presley          Roots Remember
51    Olympia   Pfeiffer             Santa Paris
52      Ellen    Presley           Scarface Bang
53      Ellen    Presley         Secretary Rouge
54    Olympia   Pfeiffer             Sense Greek
55      Ellen    Presley                Spy Mile
56    Olympia   Pfeiffer         Storm Happiness
57      Ellen    Presley    Streetcar Intentions
58    Olympia   Pfeiffer       Sweet Brotherhood
59      Ellen    Presley            Tadpole Park
60    Olympia   Pfeiffer        Titanic Boondock
61    Olympia   Pfeiffer         Tourist Pelican
62    Olympia   Pfeiffer          Traffic Hobbit
63      Ellen    Presley        Treasure Command
64      Ellen    Presley               Turn Star
65    Olympia   Pfeiffer              Wait Cider
66      Ellen    Presley            Women Dorado
67      Julia  Zellweger               Won Dares
68      Julia  Zellweger           Wyoming Storm